From 02339f2d26928c7cd6c2f7c2819d904dadb3cf36 Mon Sep 17 00:00:00 2001
From: Luwei Zhou <b45643@freescale.com>
Date: Tue, 10 Dec 2013 14:21:15 +0800
Subject: [PATCH 0621/1072] ENGR00290679 mxc: mlb Fix the MLB150 not stable
 issue in ISOC mode.

git://git.freescale.com/imx/linux-2.6-imx.git imx_3.10.17_1.0.0_beta
commit 1072f9f98e4d4da0d2fc643e65a527876f703cd2

The MLB test bench has bug when testing the ISOC mode.When we press
stop test button when completing test, MITB will cause MLB150 on ARD
error.The mlbintr ISR handler is used to handle the error interrupt.
In the ISR handler, the MLB150_MS0,MLB150MS1 should be cleared.
The spec doesn't give detailed description. The spec only says that
the registers need to be cleared before enabling interrupt.

Signed-off-by: Luwei Zhou <b45643@freescale.com>
(cherry picked from commit 567ec335491372932e13c848eb460e75eec403cb)
Signed-off-by: Hongbo Zhong <hongbo.zhong@windriver.com>
---
 drivers/mxc/mlb/mxc_mlb150.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/drivers/mxc/mlb/mxc_mlb150.c b/drivers/mxc/mlb/mxc_mlb150.c
index fe0239a..d1bae05 100755
--- a/drivers/mxc/mlb/mxc_mlb150.c
+++ b/drivers/mxc/mlb/mxc_mlb150.c
@@ -1783,6 +1783,15 @@ static irqreturn_t mlb_isr(int irq, void *dev_id)
 	 */
 	ms0 = __raw_readl(mlb_base + REG_MS0);
 	ms1 = __raw_readl(mlb_base + REG_MS1);
+
+	/*
+	 * The MLB150_MS0, MLB150_MS1 registers need to be cleared. In
+	 * the spec description, the registers should  be cleared when
+	 * enabling interrupt. In fact, we also should clear it in ISR.
+	 */
+	__raw_writel(0, mlb_base + REG_MS0);
+	__raw_writel(0, mlb_base + REG_MS1);
+
 	pr_debug("mxc_mlb150: mlb interrupt:0x%08x 0x%08x\n",
 			(u32)ms0, (u32)ms1);
 
-- 
1.7.5.4

